### DISCUSSION TITLE
PHP script to hash password

### DISCUSSION THREAD

**Discussion by FernandoGarcia** [emqx/emqx#10720] [Q&A] [answered]
Hi!

I would like to know if someone can give me a PHP script to generate the hashed password compatible with emqx.

I have this script that works fine with [mosquitto-auth-plug](https://github.com/jpmens/mosquitto-auth-plug) but I'm not able to make it work with emqx_mysql_auth plugin.

```php
<?php
define("PBKDF2_HASH_ALGORITHM", "sha256");
define("PBKDF2_ITERATIONS", 901);
define("PBKDF2_SALT_BYTE_SIZE", 12);
define("PBKDF2_HASH_BYTE_SIZE", 24);
define("SEPARATOR", "$");
define("TAG", "PBKDF2");

function create_hash($password) {
    
    $salt = base64_encode(mcrypt_create_iv(PBKDF2_SALT_BYTE_SIZE, MCRYPT_DEV_URANDOM));
    return TAG . SEPARATOR . PBKDF2_HASH_ALGORITHM . SEPARATOR . PBKDF2_ITERATIONS . SEPARATOR .  $salt . SEPARATOR . 
        base64_encode(pbkdf2(
            PBKDF2_HASH_ALGORITHM,
            $password,
            $salt,
            PBKDF2_ITERATIONS,
            PBKDF2_HASH_BYTE_SIZE,
            true
        ));
}

function pbkdf2($algorithm, $password, $salt, $count, $key_length, $raw_output = false) {
    $algorithm = strtolower($algorithm);
    if(!in_array($algorithm, hash_algos(), true))
        trigger_error('PBKDF2 ERROR: Invalid hash algorithm.', E_USER_ERROR);
    if($count <= 0 || $key_length <= 0)
        trigger_error('PBKDF2 ERROR: Invalid parameters.', E_USER_ERROR);
    if (function_exists("hash_pbkdf2")) {
        // The output length is in NIBBLES (4-bits) if $raw_output is false!
        if (!$raw_output) {
            $key_length = $key_length * 2;
        }
        return hash_pbkdf2($algorithm, $password, $salt, $count, $key_length, $raw_output);
    }
    $hash_length = strlen(hash($algorithm, "", true));
    $block_count = ceil($key_length / $hash_length);
    $output = "";
    for($i = 1; $i <= $block_count; $i++) {
        // $i encoded as 4 bytes, big endian.
        $last = $salt . pack("N", $i);
        // first iteration
        $last = $xorsum = hash_hmac($algorithm, $last, $password, true);
        // perform the other $count - 1 iterations
        for ($j = 1; $j < $count; $j++) {
            $xorsum ^= ($last = hash_hmac($algorithm, $last, $password, true));
        }
        $output .= $xorsum;
    }
    if($raw_output)
        return substr($output, 0, $key_length);
    else
        return bin2hex(substr($output, 0, $key_length));
}
?>
```

Any suggestion is appreciated.

Best regards.

**Comment by keynslug** [comment]
Hi @FernandoGarcia,

EMQX supports mysql authentication out of the box, however it does not support mosquitto's PBKDF2 flavor. Only the regular PBKDF2 is supported, with a couple of options to tune the algorithm, see documentation [[1](https://www.emqx.io/docs/en/v5.0/configuration/configuration-manual.html#password-based-mysql)] [[2](https://www.emqx.io/docs/en/v5.0/configuration/configuration-manual.html#authentication)].

If you need to support mosquitto's PBKDF2 flavor (e.g. to work with existing authn database), one option I currently see is to make a plugin hooking into `client.authenticate` hook point, implementing the algo you've provided in Erlang and largely reusing the `emqx_authn_mysql` code. Another option is to look into implementing exhook gRPC server which will handle `OnClientAuthenticate` hook the way you like, e.g. using tech you're more comfortable with.

**Comment by FernandoGarcia** [comment, marked as answer]
Hi!

Thanks for your answer.

I think will be more easy regenerate all hashed passwords in this way:

```php
<?php
$salt = base64_encode(random_bytes(12));
echo $salt;
echo "              ";
echo hash('sha256', 'ec0245b'.$salt);
?> 
```

Best regards.
